### PR TITLE
Dispatch alias instead of detach in shallow_copy_and_detach when tracing

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #176224

When autograd saves a tensor for backward, SavedVariable calls
tensor_data() → shallow_copy_and_detach(), which dispatches
aten.detach.default through Python modes. During make_fx tracing,
ProxyTorchDispatchMode captures these as detach nodes in the graph.
When the traced graph is replayed with create_graph=True for
higher-order differentiation, these detach nodes block gradient flow,
producing incorrect results.

Add PyInterpreterVTable::detach_or_alias_for_save() which checks
whether PROXY mode is active (i.e., we're tracing). If so, it
dispatches aten.alias.default (preserves gradient flow). Otherwise it
dispatches aten.detach.default (prevents reference cycles in eager
mode). Replace all detach() calls in shallow_copy_and_detach_core
implementations with the new method.
Fixes https://github.com/pytorch/pytorch/issues/175477

Authored with Claude.

### CODE REVIEW

File: test/test_proxy_tensor.py [hunk 1]
```diff
@@ -420,6 +420,39 @@ def f_backward(x):
...
+        # Autograd's save-for-backward internally detaches saved tensors.
+        # make_fx must decompose these detach ops to alias so they don't
+        # block gradient flow when the traced graph is replayed with
+        # create_graph=True for higher-order differentiation.
+        def fn(x, weight):
+            x = x.detach().requires_grad_(True)
+            h = x @ weight.T
+            energy = (h / torch.sqrt((h ** 2).mean(-1, keepdim=True) + 1e-5)).pow(2).sum()
+            force = -torch.autograd.grad(energy, x, create_graph=True)[0]
+            return energy, force
```
Comment (frgossen) [inline test/test_proxy_tensor.py:436]: I think you can reproduce this w/o the energy contribution

File: test/test_proxy_tensor.py [hunk 2]
```diff
@@ -420,6 +420,39 @@ def f_backward(x):
...
+    def test_make_fx_second_order_grad(self):
+        # Regression test for https://github.com/pytorch/pytorch/issues/175477
+        # Autograd's save-for-backward internally detaches saved tensors.
+        # make_fx must decompose these detach ops to alias so they don't
+        # block gradient flow when the traced graph is replayed with
+        # create_graph=True for higher-order differentiation.
+        def fn(x, weight):
+            x = x.detach().requires_grad_(True)
+            h = x @ weight.T
+            energy = (h / torch.sqrt((h ** 2).mean(-1, keepdim=True) + 1e-5)).pow(2).sum()
```
Comment (frgossen) [inline test/test_proxy_tensor.py]: You can already trigger this with `energy = torch.sqrt(torch.abs(h)).sum()` if you like a simpler test case.

**Review by frgossen — CHANGES_REQUESTED** [review]

File: test/test_proxy_tensor.py [hunk 3]
```diff
@@ -420,6 +420,41 @@ def f_backward(x):
...
+            force = -torch.autograd.grad(energy, x, create_graph=True)[0]
+            return energy, force
+
+        x = torch.randn(4)
+        w = torch.randn(4, requires_grad=True)
+
+        # Eager reference
+        w.grad = None
+        e_eager, f_eager = fn(x, w)
+        (e_eager + f_eager.sum()).backward()
```
Comment (frgossen) [inline test/test_proxy_tensor.py:444]: nit: I don't think you need the `energy` contribution here

**Review by frgossen — APPROVED** [review]

**Review by zou3519 — COMMENTED** [review]
This isn't a support use case for make-fx. A graph module produced from make_fx isn't supposed to be trainable. This is because it is a post-dispatch graph, and we do not expect post-dispatch graphs to be trainable.

I have not thought about if "alias instead of detach" makes sense here. I think @ezyang added the original version of this. The detach was supposed to be a "shallow_copy_and_detach" and we decided to approximate it with just a detach().

**Comment by aorenste** [comment]
Closing since it's expected not to work
